lib/sql.ml

```diff
@@ -431,6 +435,8 @@ type insert_action =
   on_duplicate : assignments option;
 }
 
+type cte = { cte_name: string;  cols: string list option; stmt: select_full; }
```

Comment:
i would expect schema type here, not string list?

Reply:
It was my first idea, but:

schema type is list of

```ocaml
type attr = {
  name : string;
  domain : Type.t;
  extra : Constraints.t;
}
```

And in this case: 

```sql
WITH RECURSIVE sequence_cte(num) AS (
    SELECT 1 AS num
    UNION ALL
    SELECT num + 1
    FROM sequence_cte
    WHERE num < 10
)
SELECT num
FROM sequence_cte;
```

cols

```ocaml
cols: string list option
```
is 

```ocaml
Some ["num"]
```

Lets look at the rest fields:  `domain : Type.t;` - we don't know a type before we check whole cte, 
`extra : Constraints.t;`- it isn't a real table

Reply:
actually i am not sure if subquery is done this way, but it would be nice - it doesn't matter if table is real or not - table is just schema with a name

Reply:
>  it doesn't matter if table is real or not - table is just schema with a name

Yes , yes, and this is how it works

```ocaml
 let s1, p1, tbls, cardinality = eval_select env (fst stmt.select) in
        let a1 = from_schema s1 in
        let ctes = if is_recursive then (tbl_name, a1) :: env.ctes else env.ctes in
        eval_compound
          ~env:{ env with tables = tbls; ctes  }
          (p1, s1, cardinality, stmt)
``` 
Here we infer types and form table of it ( schema with a name)

```ocaml
(tbl_name, s2)
```

But this cols: 

```ocaml
let select_all = Option.default (List.map (fun i -> i.name) a1) cte.cols in
```
These are just aliases, (how we expose columns (already inferred columns)), here is num for example

```sql
WITH RECURSIVE sequence_cte(num) AS (
```
 we just "glue" them. Or better to say rename columns in schema

Reply:
![explain](https://github.com/user-attachments/assets/92ad2d1a-1b41-4837-b39b-619b7f598e96)

Reply:
Done https://github.com/ahrefs/sqlgg/pull/25/commits/b3c96e322effdcdc5cc3ae3f4014945b02136a0b